### PR TITLE
Add support for debug info.

### PR DESCRIPTION
**Pull request by nlewycky**
## Description

This is built off of epilys' PR #180. I've addressed the review feedback and added the extensions that I need for my use case.

## Related Issue

#180 

## How This Has Been Tested

I've used this to generate line table info but indeed there's no tests included yet.

## Checklist

- [x] I have read the [Contributing Guide](https://github.com/TheDan64/inkwell/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: src/debug_info.rs [hunk 1]
```diff
@@ -1,30 +1,1221 @@
...
+        flags: &str,
+        runtime_ver: libc::c_uint,
+        split_name: &str,
+        kind: DWARFEmissionKind,
+        dwo_id: libc::c_uint,
+        split_debug_inlining: bool,
+        debug_info_for_profiling: bool,
+    ) -> DICompileUnit<'ctx> {
+        if self.has_cu {
+            panic!("Only one compile unit per DebugInfoBuilder.");
```
Comment (TheDan64) [inline src/debug_info.rs:194]: Does this need to panic? Shouldn't we return a Result?

File: src/debug_info.rs [hunk 2]
```diff
@@ -1,30 +1,1221 @@
...
+    pub fn create_basic_type(
+        &self,
+        name: &str,
+        size_in_bits: u64,
+        encoding: LLVMDWARFTypeEncoding,
+    ) -> DIBasicType<'ctx> {
+        if name.empty() {
+            // Also, LLVM returns the same type if you ask for the same
+            // (name, size_in_bits, encoding).
+            panic!("basic types must have names");
```
Comment (TheDan64) [inline src/debug_info.rs:357]: Result?

File: src/debug_info.rs [hunk 3]
```diff
@@ -1,30 +1,1221 @@
...
+        unsafe { LLVMMetadataReplaceAllUsesWith(placeholder.metadata_ref, other.metadata_ref) };
+    }
+
+    /// Construct any deferred debug info descriptors. May generate invalid metadata if debug info
+    /// is incomplete. Module/function verification can then fail.
+    ///
+    /// Call before any kind of code generation (including verification). Can be called more than once.
+    pub fn finalize(&self) {
+        if !self.placeholders.is_empty() {
+            panic!("Can't finalize debug info builder with outstanding placeholder types!");
```
Comment (TheDan64) [inline src/debug_info.rs:790]: Result?

File: src/debug_info.rs [hunk 4]
```diff
@@ -1,30 +1,1221 @@
...
 }
 
-impl Drop for DebugInfoBuilder {
+impl<'ctx> Drop for DebugInfoBuilder<'ctx> {
     fn drop(&mut self) {
-        // FIXME: Docs say DIB must be finalize before calling this
-        unsafe {
-            LLVMDisposeDIBuilder(self.0)
+        if !self.has_cu {
+            panic!("A DebugInfoBuilder must have a compile unit.");
```
Comment (TheDan64) [inline src/debug_info.rs:799]: Result?

File: src/debug_info.rs [hunk 5]
```diff
@@ -1,30 +1,1221 @@
...
+        &mut self,
+        placeholder: DIDerivedType<'ctx>,
+        other: DIDerivedType<'ctx>,
+    ) {
+        // `placeholder` must be in the placeholder list.
+        // `other` may or may not be a placeholder.
+        let index = self
+            .placeholders
+            .iter()
+            .position(|ph| *ph == placeholder.metadata_ref)
```
Comment (TheDan64) [inline src/debug_info.rs:778]: Does this intentionally stop at the first one? Can there be multiple?

File: src/debug_info.rs (hunk 1 above)
Comment (nlewycky) [reply src/debug_info.rs]: This panic! is removed.

File: src/debug_info.rs (hunk 2 above)
Comment (nlewycky) [reply src/debug_info.rs]: Done.

File: src/debug_info.rs (hunk 3 above)
Comment (nlewycky) [reply src/debug_info.rs]: This panic is removed.

File: src/debug_info.rs (hunk 4 above)
Comment (nlewycky) [reply src/debug_info.rs]: This panic is removed.

File: src/debug_info.rs (hunk 5 above)
Comment (nlewycky) [reply src/debug_info.rs]: It was stopping at the first one and they needed to be unique. Regardless, this entire technique doesn't work (DebugInfoBuilder is Copy and then the `placeholders` member gets out of sync) so this has been deleted.

File: src/module.rs [hunk 6]
```diff
@@ -1358,20 +1357,9 @@ impl<'ctx> Module<'ctx> {
...
-            if allow_unresolved {
-                LLVMCreateDIBuilder(self.module.get())
-            } else {
-                LLVMCreateDIBuilderDisallowUnresolved(self.module.get())
-            }
-        };
-
-        DebugInfoBuilder::new(dib)
+    #[llvm_versions(7.0..=latest)]
+    pub fn create_debug_info_builder(&self, allow_unresolved: bool) -> DebugInfoBuilder<'ctx> {
```
[the remark this answers is not in the document]
Comment (nlewycky) [reply src/module.rs]: I was going off of `create_debug_info_builder` which is marked as being 7+. I haven't actually checked llvm-sys. We could move it to 6 and see if it works on travis?

[the remark this answers is not in the document]
Comment (TheDan64) [reply src/module.rs]: https://gitlab.com/taricorp/llvm-sys.rs/-/blob/llvm-6.0/src/debuginfo.rs
https://gitlab.com/taricorp/llvm-sys.rs/-/blob/llvm-7.0/src/debuginfo.rs

It definitely won't work wholesale but would in part. Do you think it's worth supporting 6? Might not be with how little is there

[the remark this answers is not in the document]
Comment (nlewycky) [reply src/module.rs]: My general answer is that we don't need to add new features for old LLVMs, people using inkwell+llvm 7 are already happy with the features it supports and people who start using inkwell today will pick the newest llvm.

That aside. It looks like for LLVM 6 they tried to include just enough to make line tables work, except that I don't see Subprogram support. When I tried dbg! statements without declaring a subprogram in LLVM 10 the verifier complained that the scope wasn't a local scope. LLVM 6 might permit that though?

**Comment by Zaxar163** [comment]
When will this Pull Request be accepted?

**Comment by TheDan64** [comment]
@Zaxar163 Hopefully soon. We just need to figure out why LLVM 10 decided to stop linking: https://travis-ci.org/github/TheDan64/inkwell/jobs/708168713. I believe that's the only blocker

**Comment by TheDan64** [comment]
Thanks!

**Pull request merged**